### PR TITLE
Fix of loading boundary elements of ParMesh from files

### PR DESCRIPTION
**Pull request by najlkin**
This PR fixes the problem recognized in #4083 . When `ParMesh` is loaded from files/streams through its constructor or `Load()` method with `refine` or `fix_orientation` parameters set true (as by default), boundary elements are generated for every boundary face including the shared ones if the local mesh does not have by chance at least one boundary element already loaded from the file, in which case the global boundary is preserved and no boundary elements are generated. This is a completely inconsistent behavior causing spurious appearance of internal boundaries depending on partitioning of the `ParMesh`.

The problem is/was in method `Mesh::GenerateBoundaryElements()`, which does not check anyhow if the boundary is shared or not. Here, it is overridden in `ParMesh` and no boundary elements are generated on shared faces.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4128](https://github.com/mfem/mfem/pull/4128) | @najlkin | @tzanio | @sebastiangrimberg + @sohailreddy + @jandrej | 2/10/24 | 4/12/24 | 4/15/24 | |

<details>
<summary>PR Checklist</summary>
    
- [x] Code builds.
- [x] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [x] Run `make unittest` to make sure all unit tests pass.
- [x] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

File: mesh/pmesh.cpp [hunk 1]
```diff
@@ -969,6 +969,9 @@ void ParMesh::Load(istream &input, int generate_edges, int refine,
       pncmesh->GetConformingSharedStructures(*this);
    }
 
+   // Assume FinalizeTopology() has been called in Loader
+   FinalizeParTopo();
```
Comment (sebastiangrimberg) [inline mesh/pmesh.cpp:973]: `ParMesh::Finalize` already calls `FinalizeParTopo`. Is this addition required?

File: mesh/pmesh.cpp [hunk 2]
```diff
@@ -3190,6 +3193,53 @@ int ParMesh::GetNFbyType(FaceType type) const
...
+   Array<bool> l2s_marker(GetNumFaces());
+   l2s_marker = false;
+   for (int l : *s2l_face) { l2s_marker[l] = true; }
+
+   // count the 'NumOfBdrElements'
+   NumOfBdrElements = 0;
+   for (int i = 0; i < faces_info.Size(); i++)
+   {
+      const auto &fi = faces_info[i];
+      if (fi.Elem2No < 0 && !l2s_marker[i])
```
Comment (sebastiangrimberg) [inline mesh/pmesh.cpp:3224]: I think this can be simplified to check `fi.Elem2Inf`. Or just use `FaceIsTrueInterior` in the `Mesh::GenerateBoundaryElements` implementation instead of just checking `fi.Elem2No` (which is like checking `FaceIsInterior`).

File: mesh/pmesh.cpp (hunk 1 above)
Comment (najlkin) [reply mesh/pmesh.cpp:973]: Yes, it is, otherwise `sedge_ledge` and similar are not initialized when we get to `GenerateBoundaryElements()`.

File: mesh/pmesh.cpp (hunk 2 above)
Comment (najlkin) [reply mesh/pmesh.cpp:3224]: No, it does not work because this encoded information about faces is available only after `ExchangeFaceNbrData()` has been called (as the docstring of `FaceIsInterior` mentions). The faces at this point just come from `Mesh::GenerateFaces()`, so there are not many options really.

File: mesh/pmesh.cpp (hunk 1 above)
Comment (najlkin) [reply mesh/pmesh.cpp:973]: Maybe I could make it conditional if `refine` or `fix_orientations` are true, because only then the topology is possibly changing in `Finalize`, but I do not know, it might be confusing... 🤔

Comment (sebastiangrimberg) [reply mesh/pmesh.cpp:973]: Maybe it would be more clear if inside `ParMesh::Finalize` you just moved `FinalizeParTopo` to be called before `Mesh::Finalize`?

Comment (najlkin) [reply mesh/pmesh.cpp:973]: I am not sure, the topology might be changing also at other places, there is maybe a good reason for that... 🤔

Comment (najlkin) [reply mesh/pmesh.cpp:973]: Yeah, it does DOFs reordering with curved meshes, so the edge/face indices are then completely different then and `FinalizeParTopo` must be done after that.

Comment (najlkin) [reply mesh/pmesh.cpp:973]: > That's true. Another option would be inside of the `ParMesh::GenrateBoundaryElements` overload where these data structures are used. I'm just thinking of how this change could be made more clear/reduce the distance between it and the intended side effects.

I am not sure I follow, do you mean to call `FinalizeParTopo` from there? For some reason it is a public method, so a user may call it any time. So I do not think it is a good idea...

Comment (sebastiangrimberg) [reply mesh/pmesh.cpp:973]: I am suggesting to remove the added call here (as others may ask why it's needed after this PR) and replace it with a call at the start of ParMesh::GenerateBoundaryElements(), at line 3209, for example. This way it's clear why you need to call it.

Comment (najlkin) [reply mesh/pmesh.cpp:973]: Yeah, I would agree when it was a private method how it used to be, but `GenerateBoundaryElements` is public nowadays, so it is dangerous to add side effects to it, when a user may call it any time.

Comment (najlkin) [reply mesh/pmesh.cpp:973]: How about just a comment in the code? :wink:

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: mesh/pmesh.cpp (hunk 1 above)
Comment (sebastiangrimberg) [reply mesh/pmesh.cpp:973]: I think this is an improvement, since it avoids calling `FinalizeParTopo` twice unless absolutely necessary, while also keeping `GenerateBoundaryElements` fine for users to call as a public method:

```
diff --git a/mesh/pmesh.cpp b/mesh/pmesh.cpp
index 999e564c4..9b3db80c2 100644
--- a/mesh/pmesh.cpp
+++ b/mesh/pmesh.cpp
@@ -969,9 +969,6 @@ void ParMesh::Load(istream &input, int generate_edges, int refine,
       pncmesh->GetConformingSharedStructures(*this);
    }

-   // Assume FinalizeTopology() has been called in Loader
-   FinalizeParTopo();
-
    Finalize(refine, fix_orientation);

    EnsureParNodes();
@@ -3206,6 +3203,11 @@ void ParMesh::GenerateBoundaryElements()
       bel_to_edge = NULL;
    }

+   if (shared_edges.Size() != sedge_ledge.Size())
+   {
+      FinalizeParTopo();
+   }
+
    const Array<int> *s2l_face;
    if (Dim == 1) { s2l_face = &svert_lvert; }
    else if (Dim == 2) { s2l_face = &sedge_ledge; }
```

**Comment by sebastiangrimberg** [comment]
This is a good bug fix and resolves the issue in https://github.com/mfem/mfem/issues/4083. I would suggest also including the following change to `mesh.cpp` as part of this PR to avoid calling `GenerateBoundaryElements` in parallel situations when it is unintended:

```
diff --git a/mesh/mesh.cpp b/mesh/mesh.cpp
index ce8dc7cf5..b830c499b 100644
--- a/mesh/mesh.cpp
+++ b/mesh/mesh.cpp
@@ -3014,7 +3014,7 @@ void Mesh::FinalizeTetMesh(int generate_edges, int refine, bool fix_orientation)
    FinalizeCheck();
    CheckElementOrientation(fix_orientation);

-   if (NumOfBdrElements == 0)
+   if (ReduceInt(NumOfBdrElements) == 0)
    {
       GetElementToFaceTable();
       GenerateFaces();
@@ -3056,7 +3056,7 @@ void Mesh::FinalizeWedgeMesh(int generate_edges, int refine,
    FinalizeCheck();
    CheckElementOrientation(fix_orientation);

-   if (NumOfBdrElements == 0)
+   if (ReduceInt(NumOfBdrElements) == 0)
    {
       GetElementToFaceTable();
       GenerateFaces();
@@ -3093,7 +3093,7 @@ void Mesh::FinalizeHexMesh(int generate_edges, int refine, bool fix_orientation)
    GetElementToFaceTable();
    GenerateFaces();

-   if (NumOfBdrElements == 0)
+   if (ReduceInt(NumOfBdrElements) == 0)
    {
       GenerateBoundaryElements();
    }
@@ -3166,7 +3166,7 @@ void Mesh::FinalizeTopology(bool generate_bdr)
    {
       GetElementToFaceTable();
       GenerateFaces();
-      if (NumOfBdrElements == 0 && generate_bdr)
+      if (ReduceInt(NumOfBdrElements) == 0 && generate_bdr)
       {
          GenerateBoundaryElements();
          GetElementToFaceTable(); // update be_to_face
@@ -3186,7 +3186,7 @@ void Mesh::FinalizeTopology(bool generate_bdr)
       if (Dim == 2)
       {
          GenerateFaces(); // 'Faces' in 2D refers to the edges
-         if (NumOfBdrElements == 0 && generate_bdr)
+         if (ReduceInt(NumOfBdrElements) == 0 && generate_bdr)
          {
             GenerateBoundaryElements();
          }
@@ -3200,7 +3200,7 @@ void Mesh::FinalizeTopology(bool generate_bdr)
    if (Dim == 1)
    {
       GenerateFaces();
-      if (NumOfBdrElements == 0 && generate_bdr)
+      if (ReduceInt(NumOfBdrElements) == 0 && generate_bdr)
       {
          // be_to_face will be set inside GenerateBoundaryElements
          GenerateBoundaryElements();
@@ -3264,7 +3264,7 @@ void Mesh::Finalize(bool refine, bool fix_orientation)
    }
    if (refine)
    {
-      MarkForRefinement();   // may change topology!
+      MarkForRefinement(); // may change topology!
    }

    if (may_change_topology)
@@ -3277,7 +3277,8 @@ void Mesh::Finalize(bool refine, bool fix_orientation)
       }
       else
       {
-         FinalizeTopology(); // Re-computes some data unnecessarily.
+         bool generate_bdr = false;
+         FinalizeTopology(generate_bdr); // re-computes some data unnecessarily
       }

       // TODO: maybe introduce Mesh::NODE_REORDER operation and FESpace::
```

File: mesh/pmesh.cpp (hunk 1 above)
Comment (najlkin) [reply mesh/pmesh.cpp:973]: I may move `FinalizeParTopo` closer to the source to `LoadSharedEntities` and `ParNCMesh::GetConformingSharedStructures()`, where the structures like `sedge_ledge` are allocated, but not initialized, which is a bad practice by itself. However, I still cannot remove the second call from `ParMesh::Finalize()`, which might be unnecessary sometimes, but it is hard to determine when exactly. In this case of loading from files, it is unnecessary when the topology has not been changed, but there can be other scenarios when the user calls `Finalize` directly (it is a public method) and it is expected to construct these structures (and there is no other way to call `FinalizeParTopo` as it is protected).

Comment (sebastiangrimberg) [reply mesh/pmesh.cpp:973]: Indeed it seems like a bad idea that `sface_lface` and `sedge_ledge` are allocated but not filled by `LoadSharedEntities`.

OK, maybe a better suggestion is to follow the code in `ParMesh::MarkTetMeshForRefinement`. Just avoid calling `FinalizeParTopo` altogether and use the `shared_edges`, `shared_trias`, and `shared_quads` members in your `GenerateBoundaryElements` instead. See this block:

```
Array<int> sedge_ord(shared_edges.Size());
Array<Pair<int,int> > sedge_ord_map(shared_edges.Size());
for (int k = 0; k < shared_edges.Size(); k++)
{
   // sedge_ledge may be undefined -- use shared_edges and v_to_v instead
   const int sedge = group_sedge.GetJ()[k];
   const int *v = shared_edges[sedge]->GetVertices();
   sedge_ord[k] = order[v_to_v(v[0], v[1])];
}
```

This might be the safest (and most optimal in terms of not recomputing quantities) route.

Comment (najlkin) [reply mesh/pmesh.cpp:973]: Yes, that is also an option. However, I have checked other constructors and they also initialize `sedge_ledge` and I do not see any other ways how to manipulate these structures, so it is unnecessary in `Finalize` to set them and it is rather a hidden side effect. It should be safe and a good practice to move their initialization to the source (`LoadSharedEntities` and `ParNCMesh::GetConformingSharedStructures()`). I would just need to find a way how to hijack the path when the topology is changed in `Finalize`.

**Comment by najlkin** [comment]
I am not sure about `bool generate_bdr = false;`. It is an intended behavior perhaps 🤔

**Comment by sebastiangrimberg** [comment]
> I am not sure about `bool generate_bdr = false;`. It is an intended behavior perhaps 🤔

Because `Finalize` already assumes `FinalizeTopology` has been called, there are two options:
- Either `FinalizeTopology` was called with `generate_bdr = true`, in which case calling it here with `generate_bdr = false` has no effect
- `FinalizeTopology` was called with `generate_bdr = false`, in which case calling it here with `generate_bdr = true` as is currently the behavior on `master` is producing unintended consequences for the user

So, I think in either of these cases adding the explicit `generate_bdr = false` argument is not going to introduce unwelcome changes. But the other reviewers should chime in.

File: mesh/pmesh.cpp (hunk 1 above)
Comment (najlkin) [reply mesh/pmesh.cpp]: How about now? I added a protected method `UpdatedTopology()` to hijack the case with the change of topology, but it is not fully self-purposed, I believe, and might be useful in the future too.

**Comment by najlkin** [comment]
To move things forward a little bit, @jandrej , @sohailreddy , what are your opinions on `generate_bdr = false`?

**Comment by tzanio** [comment]
ping: @jandrej , @sohailreddy

File: mesh/mesh.cpp [hunk 3]
```diff
@@ -3014,7 +3014,7 @@ void Mesh::FinalizeTetMesh(int generate_edges, int refine, bool fix_orientation)
    FinalizeCheck();
    CheckElementOrientation(fix_orientation);
 
-   if (NumOfBdrElements == 0)
+   if (ReduceInt(NumOfBdrElements) == 0)
```
Comment (jandrej) [inline mesh/mesh.cpp:3017]: Can we work around this? I don't like introducing parallel paradigms into serial classes

Comment (najlkin) [reply mesh/mesh.cpp:3017]: Welcome to MFEM 😄 This is the way how compatibility is achieved between `Mesh` and `ParMesh`, so `Mesh::ReduceInt` does nothing in fact 🙃 . I do not think there is a real work around, maybe cosmetic, but global number of (boundary) elements is not stored, which would require an update every time something happens somewhere globally, which is error prone 💣 . Instead, the other approach is taken collecting the global numbers when you need them, which should be (and normally is) rare.

Comment (jandrej) [reply mesh/mesh.cpp:3017]: Just because there are occurrences in the codebase where suboptimal choices have been made doesn't mean we ignore better practice in the future.

Why can the reduce not happen at the `ParMesh::GenerateBoundaryElements` level?

Comment (najlkin) [reply mesh/mesh.cpp:3017]: And then someone tells me to keep the changes minimal and why do I make things complicated 😞 ... but that is a different story, I definitely support cleaning the code bit by bit 😉 
The thing here is that `GenerateBoundaryElements` does not check anything and overwrites the existing boundary eventually, which might be desired behavior, so it should not go there. However, we may introduce something like `EnsureBoundaryElements` (like `EnsureParNodes` or other), which will do the check and call `GenerateBoundaryElements` if needed. It may then have serial and parallel implementations (i.e. `Mesh` vs. `ParMesh`). How about that? 😉

Comment (najlkin) [reply mesh/mesh.cpp:3017]: Unfortunately, that will not help everywhere, because more things are done at some places. The only thing I see as possible is to add `GetGlobalNBE` like the existing `GetGlobalNE` 🤔 .

Comment (najlkin) [reply mesh/mesh.cpp:3017]: ...or use `HasBoundaryElements()` 🤔

Comment (najlkin) [reply mesh/mesh.cpp]: Ok, here we go, `HasBoundaryElements()` sounds as the cleanest way 👍 . Interestingly, it was not used anywhere 🙃

**Comment by tzanio** [comment]
Now that PR https://github.com/mfem/mfem/pull/3922 has been merged in `master`, please update this PR by:

1. merging `master` in here and resolving conflicts
2. replacing use of `double` with `real_t`
3. guarding double/single-only sections with `MFEM_USE_DOUBLE` and `MFEM_USE_SINGLE`
4. making sure tests run cleanly with default and `MFEM_PRECISION=single` configurations

See https://github.com/orgs/mfem/discussions/4207 for more details and/or ask @dylan-copeland for help.

File: mesh/pmesh.cpp (hunk 1 above)
Comment (najlkin) [reply mesh/pmesh.cpp]: ping @sebastiangrimberg

File: mesh/mesh.cpp (hunk 3 above)
Comment (najlkin) [reply mesh/mesh.cpp]: ping @jandrej

**Comment by sebastiangrimberg** [comment]
I'm still not fully convinced by this solution, and wonder if it's even necessary to support generating boundary elements for parallel mesh objects (why wouldn't you just do it for the serial mesh before partitioning?). In that case I would support just defining the `ParMesh::GenerateBoundaryElements` to be an empty function `{ }` or to throw an error to the user. I think the added complexities here about when things are initialized/invalidated is bound to cause problems down the road.

The changes to to checks for `NumOfBdrElements == 0` to `!HasBoundaryElements()` are good though.

**Comment by najlkin** [comment]
> I'm still not fully convinced by this solution, and wonder if it's even necessary to support generating boundary elements for parallel mesh objects (why wouldn't you just do it for the serial mesh before partitioning?). In that case I would support just defining the `ParMesh::GenerateBoundaryElements` to be an empty function `{ }` or to throw an error to the user. I think the added complexities here about when things are initialized/invalidated is bound to cause problems down the road.

When we opened this question at one of the meetings, the prevailing opinion seemed to be to keep consistency between serial and parallel implementations and keep backward compatibility, so I would adhere to that 😉 .
 
> The changes to to checks for `NumOfBdrElements == 0` to `!HasBoundaryElements()` are good though. Since it involves a global reduction though, I would just call it once in the functions like `FinalizeTopology` and store the value rather than using it repeatedly, if possible.

That `HasBoundaryElements()` is not called multiple times, just once normally. It appears in methods for different meshes `Finalize*Mesh`. In `FinalizeTopology()`, it is called on one of the paths depending on the dimension. It is more a question of duplicity of the code, which is bad in `Mesh`, but that goes beyond this PR 😐

**Comment by najlkin** [comment]
Ok, as discussed at the meeting, I am reverting that move of the initialization of shared structures, which was meant to make the code cleaner and better maintainable. So it will go to a separate PR and this one is just the basic fix of the loading without any bells and whistles 🔕 😉 .

**Comment by sebastiangrimberg** [comment]
My only complaint is that it still is unfortunate that `ParMesh::Load` will now call `FinalizeParTopo` twice always (once in `Load`, once in `Finalize`). This is required for the (rare) possibility of needing to access the parallel data structures in the `ParMesh::GenerateBoundaryElements` overload, and does seem like something worth avoiding.

**Comment by najlkin** [comment]
Yes, that is where we have started 😄 . I edited the header of #4240 to mention it avoids this as well 😉 . Please keep this PR minimal, so it can get finally merged. The overhead is only in the loader, so practically negligible and not caused by this PR.

File: mesh/pmesh.cpp (hunk 1 above)
Comment (najlkin) [reply mesh/pmesh.cpp:973]: Moved to #4240 .

**Comment by sebastiangrimberg** [comment]
> Yes, that is where we have started 😄 . I edited the header of #4240 to mention it avoids this as well 😉 . Please keep this PR minimal, so it can get finally merged. The overhead is only in the loader, so practically negligible and not caused by this PR.

Just need to reemphasize that my suggestion to remove the boundary element generation in the first part of [this comment](https://github.com/mfem/mfem/pull/4128#issuecomment-2037631281) would be one way of remove that limitation and make the fixes in this PR even more minimal. I don't think there was clear consensus that this was a necessary feature to add though maybe I am missing something.

Another option would be to use the approach selected in [this comment](https://github.com/mfem/mfem/pull/4128#discussion_r1486631185) and avoid accessing `sedge_ledge` or `sface_lface` inside of `ParMesh::GenerateBoundaryElements`. You can get the same information from the (already constructed) `shared_edges`, `shared_trias`, and `shared_quads` for your boundary element generation.

**Comment by najlkin** [comment]
> Just need to reemphasize that my suggestion to remove the boundary element generation in the first part of https://github.com/mfem/mfem/pull/4128#issuecomment-2037631281 would be one way of remove that limitation and make the fixes in this PR even more minimal. I don't think there was clear consensus that this was a necessary feature to add though maybe I am missing something.

This bugfix is not adding a feature, but recovers the intended behavior, which is consistent with the behavior of serial meshes. I believe majority is on the side of keeping things consistent, or at least for a bugfix. We may return to this discussion in another PR if you want to open it for this.

> Another option would be to use the approach suggested in https://github.com/mfem/mfem/pull/4128#discussion_r1486631185 and avoid accessing sedge_ledge or sface_lface inside of ParMesh::GenerateBoundaryElements. You can get the same information from the (already constructed) shared_edges, shared_trias, and shared_quads for your boundary element generation.

If we do not want to go beyond the bugfix and touch calls of `FinalizeParTopo()`, there is no need for the change. We would be just risking breaking the implementation. However, it is a good suggestion, just I am not convinced it is worth of it, but we may discuss it within #4240 😉 .

**Comment by sebastiangrimberg** [comment]
OK, I've voiced my concerns but can only defer to editors/other reviewers at this point.

**Comment by v-dobrev** [comment]
@najlkin, I think the decision at the meeting was to NOT include the parallel boundary generation but just make the checks `(NumOfBdrElements == 0)` global and leave the parallel version of `GenerateBoundary` empty, at least for now.

This way we postpone the controversial changes for further discussion but get the important bugfix in.

**Comment by najlkin** [comment]
@v-dobrev , I thought the decision was to split this PR, as I did, moving the potentially controversial change of the initialization of some structures to #4240 . The implementation of `Mesh::GenerateBoundaryElements()` in `ParMesh` is not controversial in the sense we all believe it is working. It has been tested on the code of the user and the user also confirmed it is working for him in #4083 . @sebastiangrimberg is proposing two other things:

- The alternative of removing the functionality of generating boundaries all together for `ParMesh`.
- Cleaning up the pipeline of `(Par)Mesh`, which goes along the lines of #4240 , but he would go even further.

Both are API changes and should be discussed in #4240 , but not here (!). Please someone just confirm this is working, so it can go to v4.7 asap :pray: .

**Comment by najlkin** [comment]
@jandrej , have you tested this PR is working? That is all we need now, the discussion related to the API changes should go to #4240 .

**Comment by tzanio** [comment]
As a compromise: is it just possible to just `ParMesh::GenerateBoundaryElements = delete` which will explicitly expose this as a bug to the users?

**Comment by sebastiangrimberg** [comment]
> As a compromise: is it just possible to just `ParMesh::GenerateBoundaryElements = delete` which will explicitly expose this as a bug to the users?
> 
> @sebastiangrimberg, @v-dobrev and @najlkin -- what do you think?

I like this idea but am not sure you can do that in C++ if I understand correctly:

```
./mfem/linalg/../fem/../mesh/pmesh.hpp:402:9: error: deleted function 'GenerateBoundaryElements' cannot override a non-deleted function
  402 |    void GenerateBoundaryElements() override = delete;
```

**Comment by najlkin** [comment]
@sebastiangrimberg , it must be without `virtual` (and `override`)

**Comment by sebastiangrimberg** [comment]
> @sebastiangrimberg , it must be without `virtual` (and `override`)

Oh I see, but doesn't that mean you could get this behavior for a valid `ParMesh`?

```
ParMesh pmesh(...);
pmesh.GenerateBoundaryElements();  // Compiler error
Mesh *mesh = &pmesh;
mesh->GenerateBoundaryElements();  // Not a compiler error, bug-prone?
```

**Comment by najlkin** [comment]
I think that behavior is correct in the sense that when you explicitly use the local sub-mesh, you can do many harmful things to the distributed mesh for sure, but that is responsibility of the user and might be desired for some master hacker 🤓

**Comment by sebastiangrimberg** [comment]
I think it means that when `Mesh::FinalizeTopology` eventually gets called from `ParMesh::Load`, you will get a call to `Mesh::GenerateBoundaryElements` which is what I thought was what we wanted to avoid.

**Comment by tzanio** [comment]
OK, then I vote to implement it as `{ MFEM_ABORT(" ..."); }` and note that users shouldn't be calling that function.

**Comment by najlkin** [comment]
Hmm, well, it is not critical, because that happens only when you do not have any boundary elements globally, which was not the case for the user and most other users. Anyway, I think a working implementation would be a better option, but to follow this compromise we may make `FinalizeTopology()` virtual and call the serial implementation with `generate_bdr=false` from `ParMesh`.

**Comment by najlkin** [comment]
...or abort, if `generate_bdr==true`

**Comment by najlkin** [comment]
Ok, so followed followed the pathway of @tzanio 's [comment](https://github.com/mfem/mfem/pull/4128#issuecomment-2050703154) 😉 .

**Review by sebastiangrimberg — APPROVED** [review]

**Review by tzanio — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...
